File: android/src/ui/kotlin/com/burhanrabbani/acs_flutter_sdk/AcsUiLibraryPlugin.kt
package com.burhanrabbani.acs_flutter_sdk

import android.app.Activity
import android.content.Context
import android.view.View
import android.util.Log
import androidx.annotation.NonNull
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import com.azure.android.communication.common.CommunicationTokenCredential
import com.azure.android.communication.common.CommunicationUserIdentifier
import com.azure.android.communication.ui.calling.CallComposite
import com.azure.android.communication.ui.calling.CallCompositeBuilder
import com.azure.android.communication.ui.calling.models.CallCompositeButtonViewData
import com.azure.android.communication.ui.calling.models.CallCompositeCallScreenControlBarOptions
import com.azure.android.communication.ui.calling.models.CallCompositeCallScreenOptions
import com.azure.android.communication.ui.calling.models.CallCompositeCallStateCode
import com.azure.android.communication.ui.calling.models.CallCompositeGroupCallLocator
import com.azure.android.communication.ui.calling.models.CallCompositeLocalizationOptions
import com.azure.android.communication.ui.calling.models.CallCompositeMultitaskingOptions
import com.azure.android.communication.ui.calling.models.CallCompositeRoomLocator
import com.azure.android.communication.ui.calling.models.CallCompositeSupportedScreenOrientation
import com.azure.android.communication.ui.calling.models.CallCompositeTeamsMeetingIdLocator
import com.azure.android.communication.ui.calling.models.CallCompositeTeamsMeetingLinkLocator
import com.azure.android.communication.ui.calling.models.CallCompositeErrorCode
import com.azure.android.communication.ui.calling.models.CallCompositeLocalOptions
import java.util.Locale
import java.util.UUID

/**
 * Flutter plugin for Azure Communication Services UI Library
 *
 * This plugin provides access to pre-built UI composites (CallComposite, ChatComposite)
 * from the Azure Communication Services UI Library.
 */
class AcsUiLibraryPlugin : FlutterPlugin, MethodCallHandler, ActivityAware {
    private lateinit var channel: MethodChannel
    private lateinit var context: Context
    private var activity: Activity? = null
    private var callComposite: CallComposite? = null

    companion object {
        private const val CHANNEL_NAME = "acs_ui_library"
    }

    override fun onAttachedToEngine(@NonNull binding: FlutterPlugin.FlutterPluginBinding) {
        channel = MethodChannel(binding.binaryMessenger, CHANNEL_NAME)
        channel.setMethodCallHandler(this)
        context = binding.applicationContext
    }

    override fun onDetachedFromEngine(@NonNull binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
    }

    override fun onAttachedToActivity(binding: ActivityPluginBinding) {
        activity = binding.activity
    }

    override fun onDetachedFromActivityForConfigChanges() {
        activity = null
    }

    override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {
        activity = binding.activity
    }

    override fun onDetachedFromActivity() {
        activity = null
    }

    override fun onMethodCall(@NonNull call: MethodCall, @NonNull result: Result) {
        when (call.method) {
            "launchGroupCall" -> launchGroupCall(call, result)
            "launchTeamsMeeting" -> launchTeamsMeeting(call, result)
            "launchTeamsMeetingWithId" -> launchTeamsMeetingWithId(call, result)
            "launchRoomCall" -> launchRoomCall(call, result)
            "launchOutgoingCall" -> launchOutgoingCall(call, result)
            "launchChat" -> launchChat(call, result)
            "dismiss" -> dismiss(result)
            "bringToForeground" -> bringToForeground(result)
            else -> result.notImplemented()
        }
    }

    private fun launchGroupCall(call: MethodCall, result: Result) {
        try {
            val accessToken = call.argument<String>("accessToken")
                ?: return result.error("INVALID_ARGS", "accessToken is required", null)
            val groupId = call.argument<String>("groupId")
                ?: return result.error("INVALID_ARGS", "groupId is required", null)
            val options = call.argument<Map<String, Any?>>("options")
                ?: return result.error("INVALID_ARGS", "options is required", null)

            val credential = CommunicationTokenCredential(accessToken)
            val (composite, localOptions) = buildCallComposite(credential, options)
            val locator = CallCompositeGroupCallLocator(UUID.fromString(groupId))

            activity?.let {
                composite.launch(it, locator, localOptions)
                callComposite = composite
                result.success(null)
            } ?: result.error("NO_ACTIVITY", "Activity not available", null)
        } catch (e: Exception) {
            result.error("LAUNCH_FAILED", e.message, e.stackTraceToString())
        }
    }

    private fun launchTeamsMeeting(call: MethodCall, result: Result) {
        try {
            val accessToken = call.argument<String>("accessToken")
                ?: return result.error("INVALID_ARGS", "accessToken is required", null)
            val meetingLink = call.argument<String>("meetingLink")
                ?: return result.error("INVALID_ARGS", "meetingLink is required", null)
            val options = call.argument<Map<String, Any?>>("options")
                ?: return result.error("INVALID_ARGS", "options is required", null)

            val credential = CommunicationTokenCredential(accessToken)
            val (composite, localOptions) = buildCallComposite(credential, options)
            val locator = CallCompositeTeamsMeetingLinkLocator(meetingLink)

            activity?.let {
                composite.launch(it, locator, localOptions)
                callComposite = composite
                result.success(null)
            } ?: result.error("NO_ACTIVITY", "Activity not available", null)
        } catch (e: Exception) {
            result.error("LAUNCH_FAILED", e.message, e.stackTraceToString())
        }
    }

    private fun launchTeamsMeetingWithId(call: MethodCall, result: Result) {
        try {
            val accessToken = call.argument<String>("accessToken")
                ?: return result.error("INVALID_ARGS", "accessToken is required", null)
            val meetingId = call.argument<String>("meetingId")
                ?: return result.error("INVALID_ARGS", "meetingId is required", null)
            val meetingPasscode = call.argument<String>("meetingPasscode")
                ?: return result.error("INVALID_ARGS", "meetingPasscode is required", null)
            val options = call.argument<Map<String, Any?>>("options")
                ?: return result.error("INVALID_ARGS", "options is required", null)

            val credential = CommunicationTokenCredential(accessToken)
            val (composite, localOptions) = buildCallComposite(credential, options)
            val locator = CallCompositeTeamsMeetingIdLocator(meetingId, meetingPasscode)

            activity?.let {
                composite.launch(it, locator, localOptions)
                callComposite = composite
                result.success(null)
            } ?: result.error("NO_ACTIVITY", "Activity not available", null)
        } catch (e: Exception) {
            result.error("LAUNCH_FAILED", e.message, e.stackTraceToString())
        }
    }

    private fun launchRoomCall(call: MethodCall, result: Result) {
        try {
            val accessToken = call.argument<String>("accessToken")
                ?: return result.error("INVALID_ARGS", "accessToken is required", null)
            val roomId = call.argument<String>("roomId")
                ?: return result.error("INVALID_ARGS", "roomId is required", null)
            val options = call.argument<Map<String, Any?>>("options")
                ?: return result.error("INVALID_ARGS", "options is required", null)

            val credential = CommunicationTokenCredential(accessToken)
            val (composite, localOptions) = buildCallComposite(credential, options)
            val locator = CallCompositeRoomLocator(roomId)

            activity?.let {
                composite.launch(it, locator, localOptions)
                callComposite = composite
                result.success(null)
            } ?: result.error("NO_ACTIVITY", "Activity not available", null)
        } catch (e: Exception) {
            result.error("LAUNCH_FAILED", e.message, e.stackTraceToString())
        }
    }

    private fun launchOutgoingCall(call: MethodCall, result: Result) {
        try {
            val accessToken = call.argument<String>("accessToken")
                ?: return result.error("INVALID_ARGS", "accessToken is required", null)
            val participantIds = call.argument<List<String>>("participantIds")
                ?: return result.error("INVALID_ARGS", "participantIds is required", null)
            val options = call.argument<Map<String, Any?>>("options")
                ?: return result.error("INVALID_ARGS", "options is required", null)

            val credential = CommunicationTokenCredential(accessToken)
            val (composite, localOptions) = buildCallComposite(credential, options)

            val participants = participantIds.map {
                CommunicationUserIdentifier(it)
            }

            activity?.let {
                composite.launch(it, participants, localOptions)
                callComposite = composite
                result.success(null)
            } ?: result.error("NO_ACTIVITY", "Activity not available", null)
        } catch (e: Exception) {
            result.error("LAUNCH_FAILED", e.message, e.stackTraceToString())
        }
    }

    private fun launchChat(call: MethodCall, result: Result) {
        try {
            result.error("NOT_IMPLEMENTED", "ChatComposite is not available in this build.", null)
        } catch (e: Exception) {
            result.error("LAUNCH_FAILED", e.message, e.stackTraceToString())
        }
    }

    private fun dismiss(result: Result) {
        try {
            callComposite?.dismiss()
            callComposite = null
            result.success(null)
        } catch (e: Exception) {
            result.error("DISMISS_FAILED", e.message, null)
        }
    }

    /**
     * Re-displays a composite that was minimized (sent to background) via
     * multitasking. Unlike [dismiss], this keeps the call alive and does NOT null
     * [callComposite] — it only brings the existing composite Activity back to the
     * foreground so the user returns to the live call. No-op when no composite is
     * active (e.g. the call already ended).
     */
    private fun bringToForeground(result: Result) {
        try {
            callComposite?.bringToForeground(context)
            result.success(null)
        } catch (e: Exception) {
            result.error("BRING_TO_FOREGROUND_FAILED", e.message, null)
        }
    }

    /**
     * Builds a CallComposite instance with the provided options
     */
    private fun buildCallComposite(
        credential: CommunicationTokenCredential,
        options: Map<String, Any?>
    ): Pair<CallComposite, CallCompositeLocalOptions> {
        val displayName = options["displayName"] as? String ?: "User"

        val builder = CallCompositeBuilder()
            .applicationContext(context)
            .credential(credential)
            .displayName(displayName)

        val multitaskingOptions = options["multitasking"] as? Map<String, Any?>
        if (multitaskingOptions != null) {
            val enableMultitasking = multitaskingOptions["enableMultitasking"] as? Boolean ?: true
            val enablePiP = multitaskingOptions["enablePictureInPicture"] as? Boolean ?: true

            builder.multitasking(
                CallCompositeMultitaskingOptions(enableMultitasking, enablePiP)
            )
        }

        val localizationOptions = options["localization"] as? Map<String, Any?>
        if (localizationOptions != null) {
            val languageCode = localizationOptions["languageCode"] as? String
            val countryCode = localizationOptions["countryCode"] as? String
            val isRtl = localizationOptions["isRightToLeft"] as? Boolean ?: false

            if (languageCode != null) {
                val locale = if (!countryCode.isNullOrBlank()) {
                    Locale(languageCode, countryCode)
                } else {
                    Locale(languageCode)
                }
                val layoutDirection = if (isRtl) {
                    View.LAYOUT_DIRECTION_RTL
                } else {
                    View.LAYOUT_DIRECTION_LTR
                }
                builder.localization(CallCompositeLocalizationOptions(locale, layoutDirection))
            }
        }

        val orientationOption = options["orientation"] as? String
        orientationOption?.let { orientationName ->
            val orientation = when (orientationName) {
                "portrait" -> CallCompositeSupportedScreenOrientation.PORTRAIT
                "landscape" -> CallCompositeSupportedScreenOrientation.LANDSCAPE
                "landscapeRight" -> CallCompositeSupportedScreenOrientation.USER_LANDSCAPE
                "landscapeLeft" -> CallCompositeSupportedScreenOrientation.REVERSE_LANDSCAPE
                "allButUpsideDown" -> CallCompositeSupportedScreenOrientation.FULL_SENSOR
                else -> CallCompositeSupportedScreenOrientation.FULL_SENSOR
            }
            builder.callScreenOrientation(orientation)
            builder.setupScreenOrientation(orientation)
        }

        val enableCameraButton = options["enableCameraButton"] as? Boolean ?: true
        val enableMicrophoneButton = options["enableMicrophoneButton"] as? Boolean ?: true

        val controlBarOptions = CallCompositeCallScreenControlBarOptions()
            .setCameraButton(CallCompositeButtonViewData().setEnabled(enableCameraButton))
            .setMicrophoneButton(CallCompositeButtonViewData().setEnabled(enableMicrophoneButton))

        builder.callScreenOptions(
            CallCompositeCallScreenOptions().setControlBarOptions(controlBarOptions)
        )

        val skipSetupScreen = options["skipSetupScreen"] as? Boolean ?: false
        val cameraOn = options["cameraOn"] as? Boolean ?: false
        val microphoneOn = options["microphoneOn"] as? Boolean ?: false

        val localOptions = CallCompositeLocalOptions()
            .setSkipSetupScreen(skipSetupScreen)
            .setCameraOn(cameraOn)
            .setMicrophoneOn(microphoneOn)
            .setCallScreenOptions(controlBarOptions.let { CallCompositeCallScreenOptions().setControlBarOptions(it) })

        val composite = builder.build()

        composite.addOnErrorEventHandler { errorEvent ->
            val errorCode = when (errorEvent.errorCode) {
                CallCompositeErrorCode.CALL_JOIN_FAILED -> "callJoinFailed"
                CallCompositeErrorCode.CALL_END_FAILED -> "callEndFailed"
                CallCompositeErrorCode.TOKEN_EXPIRED -> "tokenExpired"
                CallCompositeErrorCode.CAMERA_FAILURE -> "cameraFailure"
                CallCompositeErrorCode.MICROPHONE_PERMISSION_NOT_GRANTED -> "microphonePermissionNotGranted"
                CallCompositeErrorCode.NETWORK_CONNECTION_NOT_AVAILABLE -> "networkConnectionNotAvailable"
                else -> "unknown"
            }

            Log.e(
                "AcsUiLibraryPlugin",
                "onError code=$errorCode native=${errorEvent.errorCode} cause=${errorEvent.cause?.message}"
            )

            channel.invokeMethod(
                "onError",
                mapOf(
                    "errorCode" to errorCode,
                    "message" to errorEvent.cause?.message,
                    "nativeCode" to errorEvent.errorCode.toString()
                )
            )
        }

        composite.addOnCallStateChangedEventHandler { callStateEvent ->
            val state = when (callStateEvent.code) {
                CallCompositeCallStateCode.NONE -> "none"
                CallCompositeCallStateCode.CONNECTING -> "connecting"
                CallCompositeCallStateCode.RINGING -> "ringing"
                CallCompositeCallStateCode.CONNECTED -> "connected"
                CallCompositeCallStateCode.LOCAL_HOLD -> "localHold"
                CallCompositeCallStateCode.REMOTE_HOLD -> "remoteHold"
                CallCompositeCallStateCode.DISCONNECTING -> "disconnecting"
                CallCompositeCallStateCode.DISCONNECTED -> "disconnected"
                CallCompositeCallStateCode.IN_LOBBY -> "inLobby"
                else -> "unknown"
            }

            Log.d(
                "AcsUiLibraryPlugin",
                "callStateChanged native=${callStateEvent.code} mapped=$state"
            )

            channel.invokeMethod(
                "onCallStateChanged",
                mapOf(
                    "state" to state,
                    "nativeState" to callStateEvent.code.toString()
                )
            )
        }

        composite.addOnDismissedEventHandler { dismissedEvent ->
            channel.invokeMethod(
                "onDismissed",
                mapOf(
                    "errorCode" to dismissedEvent.errorCode?.toString(),
                    "reason" to dismissedEvent.cause?.message
                )
            )
            callComposite = null
        }

        composite.addOnRemoteParticipantJoinedEventHandler { participantEvent ->
            val ids = participantEvent.identifiers.map { it.rawId ?: "" }
            channel.invokeMethod(
                "onRemoteParticipantJoined",
                mapOf(
                    "participantCount" to ids.size,
                    "participantIds" to ids
                )
            )
        }

        return Pair(composite, localOptions)
    }

    private fun setupChatEventHandlers() { /* ChatComposite not implemented */ }
}

File: ios/Classes/AcsUiLibraryPlugin.swift
import Flutter
import UIKit
import AzureCommunicationCommon
import AzureCommunicationUICalling
import SwiftUI

/**
 * Flutter plugin for Azure Communication Services UI Library (iOS)
 *
 * This plugin provides access to pre-built UI composites (CallComposite)
 * from the Azure Communication Services UI Library.
 */
public class AcsUiLibraryPlugin: NSObject, FlutterPlugin {
    private var channel: FlutterMethodChannel?
    private var callComposite: CallComposite?

    private func logError(_ context: String, _ error: Error) {
        #if DEBUG
        NSLog("[AcsUiLibraryPlugin][\(context)] \(error.localizedDescription)")
        #endif
    }

    private func safeHandle(_ context: String, result: FlutterResult? = nil, _ block: () throws -> Void) {
        do {
            try block()
        } catch {
            logError(context, error)
            result?(FlutterError(code: "UNEXPECTED_ERROR", message: "Error in \(context): \(error.localizedDescription)", details: nil))
        }
    }

    public static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(
            name: "acs_ui_library",
            binaryMessenger: registrar.messenger()
        )
        let instance = AcsUiLibraryPlugin()
        instance.channel = channel
        registrar.addMethodCallDelegate(instance, channel: channel)
    }

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        safeHandle("handle:\(call.method)", result: result) {
            switch call.method {
            case "launchGroupCall":
                launchGroupCall(call: call, result: result)
            case "launchTeamsMeeting":
                launchTeamsMeeting(call: call, result: result)
            case "launchTeamsMeetingWithId":
                launchTeamsMeetingWithId(call: call, result: result)
            case "launchRoomCall":
                launchRoomCall(call: call, result: result)
            case "launchOutgoingCall":
                launchOutgoingCall(call: call, result: result)
            case "launchChat":
                launchChat(call: call, result: result)
            case "dismiss":
                dismiss(result: result)
            case "bringToForeground":
                bringToForeground(result: result)
            default:
                result(FlutterMethodNotImplemented)
            }
        }
    }

    // MARK: - CallComposite Launch Methods

    private func launchGroupCall(call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let accessToken = args["accessToken"] as? String,
              let groupId = args["groupId"] as? String,
              let options = args["options"] as? [String: Any] else {
            result(FlutterError(code: "INVALID_ARGS", message: "Missing required arguments", details: nil))
            return
        }

        do {
            let credential = try CommunicationTokenCredential(token: accessToken)
            let (composite, localOptions) = try buildCallComposite(credential: credential, options: options)

            guard let uuid = UUID(uuidString: groupId) else {
                result(FlutterError(code: "INVALID_ARGS", message: "Invalid group ID format", details: nil))
                return
            }

            DispatchQueue.main.async { [weak self] in
                composite.launch(locator: .groupCall(groupId: uuid), localOptions: localOptions)
                self?.callComposite = composite
                result(nil)
            }
        } catch {
            result(FlutterError(code: "LAUNCH_FAILED", message: error.localizedDescription, details: nil))
        }
    }

    private func launchTeamsMeeting(call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let accessToken = args["accessToken"] as? String,
              let meetingLink = args["meetingLink"] as? String,
              let options = args["options"] as? [String: Any] else {
            result(FlutterError(code: "INVALID_ARGS", message: "Missing required arguments", details: nil))
            return
        }

        #if DEBUG
        let tokenLength = accessToken.count
        NSLog("[AcsUiLibraryPlugin] launchTeamsMeeting tokenLength=\(tokenLength) meetingLink=\(meetingLink.prefix(120))")
        #endif

        do {
            let credential = try CommunicationTokenCredential(token: accessToken)
            let (composite, localOptions) = try buildCallComposite(credential: credential, options: options)

            DispatchQueue.main.async { [weak self] in
                do {
                    composite.launch(locator: .teamsMeeting(teamsLink: meetingLink), localOptions: localOptions)
                    self?.callComposite = composite
                    result(nil)
                } catch {
                    let nsError = error as NSError
                    #if DEBUG
                    NSLog("[AcsUiLibraryPlugin] launchTeamsMeeting failed domain=\(nsError.domain) code=\(nsError.code) message=\(nsError.localizedDescription)")
                    #endif
                    result(FlutterError(code: "LAUNCH_FAILED", message: nsError.localizedDescription, details: [
                        "domain": nsError.domain,
                        "code": nsError.code
                    ]))
                }
            }
        } catch {
            let nsError = error as NSError
            #if DEBUG
            NSLog("[AcsUiLibraryPlugin] launchTeamsMeeting failed during setup domain=\(nsError.domain) code=\(nsError.code) message=\(nsError.localizedDescription)")
            #endif
            result(FlutterError(code: "LAUNCH_FAILED", message: nsError.localizedDescription, details: [
                "domain": nsError.domain,
                "code": nsError.code
            ]))
        }
    }

    private func launchTeamsMeetingWithId(call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let accessToken = args["accessToken"] as? String,
              let meetingId = args["meetingId"] as? String,
              let meetingPasscode = args["meetingPasscode"] as? String,
              let options = args["options"] as? [String: Any] else {
            result(FlutterError(code: "INVALID_ARGS", message: "Missing required arguments", details: nil))
            return
        }

        do {
            let credential = try CommunicationTokenCredential(token: accessToken)
            let (composite, localOptions) = try buildCallComposite(credential: credential, options: options)

            DispatchQueue.main.async { [weak self] in
                composite.launch(
                    locator: .teamsMeetingId(
                        meetingId: meetingId,
                        meetingPasscode: meetingPasscode
                    ),
                    localOptions: localOptions
                )
                self?.callComposite = composite
                result(nil)
            }
        } catch {
            result(FlutterError(code: "LAUNCH_FAILED", message: error.localizedDescription, details: nil))
        }
    }

    private func launchRoomCall(call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let accessToken = args["accessToken"] as? String,
              let roomId = args["roomId"] as? String,
              let options = args["options"] as? [String: Any] else {
            result(FlutterError(code: "INVALID_ARGS", message: "Missing required arguments", details: nil))
            return
        }

        do {
            let credential = try CommunicationTokenCredential(token: accessToken)
            let (composite, localOptions) = try buildCallComposite(credential: credential, options: options)

            DispatchQueue.main.async { [weak self] in
                composite.launch(locator: .roomCall(roomId: roomId), localOptions: localOptions)
                self?.callComposite = composite
                result(nil)
            }
        } catch {
            result(FlutterError(code: "LAUNCH_FAILED", message: error.localizedDescription, details: nil))
        }
    }

    private func launchOutgoingCall(call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let accessToken = args["accessToken"] as? String,
              let participantIds = args["participantIds"] as? [String],
              let options = args["options"] as? [String: Any] else {
            result(FlutterError(code: "INVALID_ARGS", message: "Missing required arguments", details: nil))
            return
        }

        do {
            let credential = try CommunicationTokenCredential(token: accessToken)
            let (composite, localOptions) = try buildCallComposite(credential: credential, options: options)

            let participants = participantIds.map {
                CommunicationUserIdentifier($0)
            }

            DispatchQueue.main.async { [weak self] in
                composite.launch(participants: participants, localOptions: localOptions)
                self?.callComposite = composite
                result(nil)
            }
        } catch {
            result(FlutterError(code: "LAUNCH_FAILED", message: error.localizedDescription, details: nil))
        }
    }

    private func launchChat(call: FlutterMethodCall, result: @escaping FlutterResult) {
        result(FlutterError(
            code: "NOT_IMPLEMENTED",
            message: "ChatComposite is not available for iOS. Use custom UI with AcsChatClient.",
            details: nil
        ))
    }

    private func dismiss(result: @escaping FlutterResult) {
        DispatchQueue.main.async { [weak self] in
            self?.callComposite?.dismiss()
            self?.callComposite = nil
            result(nil)
        }
    }

    /// Re-presents a composite that was minimized (hidden) via multitasking.
    ///
    /// Unlike `dismiss`, this keeps the call alive and does NOT null the
    /// `callComposite` reference — it only restores the hidden UI. No-op when no
    /// composite is active (e.g. the call already ended).
    private func bringToForeground(result: @escaping FlutterResult) {
        DispatchQueue.main.async { [weak self] in
            // `displayCallCompositeIfWasHidden()` is private in the SDK; the public
            // entry point is the `isHidden` property — setting it to false re-shows
            // a composite that was minimized to PiP/background, keeping the call alive.
            self?.callComposite?.isHidden = false
            result(nil)
        }
    }

    // MARK: - CallComposite Builder

    private func buildCallComposite(
        credential: CommunicationTokenCredential,
        options: [String: Any]
    ) throws -> (CallComposite, LocalOptions?) {
        let displayName = options["displayName"] as? String

        let themeOptions = buildThemeOptions(options["theme"] as? [String: Any])
        let localizationOptions = buildLocalizationOptions(options["localization"] as? [String: Any])
        let orientationOption = buildOrientationOption(options["orientation"] as? String)

        let multitaskingOptions = options["multitasking"] as? [String: Any]
        let enableMultitasking = multitaskingOptions?["enableMultitasking"] as? Bool ?? false
        let enablePiP = multitaskingOptions?["enablePictureInPicture"] as? Bool ?? false

        let compositeOptions = CallCompositeOptions(
            theme: themeOptions,
            localization: localizationOptions,
            setupScreenOrientation: orientationOption,
            callingScreenOrientation: orientationOption,
            enableMultitasking: enableMultitasking,
            enableSystemPictureInPictureWhenMultitasking: enablePiP,
            displayName: displayName
        )

        let composite = CallComposite(credential: credential, withOptions: compositeOptions)

        let skipSetupScreen = options["skipSetupScreen"] as? Bool ?? false
        let cameraOn = options["cameraOn"] as? Bool ?? false
        let microphoneOn = options["microphoneOn"] as? Bool ?? false

        let enableCameraButton = options["enableCameraButton"] as? Bool ?? true
        let enableMicrophoneButton = options["enableMicrophoneButton"] as? Bool ?? true

        var callScreenOptions: CallScreenOptions?
        if !enableCameraButton || !enableMicrophoneButton {
            let controlBarOptions = CallScreenControlBarOptions(
                cameraButton: ButtonViewData(enabled: enableCameraButton),
                microphoneButton: ButtonViewData(enabled: enableMicrophoneButton)
            )
            callScreenOptions = CallScreenOptions(controlBarOptions: controlBarOptions)
        }

        let localOptions = LocalOptions(
            cameraOn: cameraOn,
            microphoneOn: microphoneOn,
            skipSetupScreen: skipSetupScreen,
            callScreenOptions: callScreenOptions
        )

        setupEventHandlers(composite: composite)

        return (composite, localOptions)
    }

    private func buildThemeOptions(_ theme: [String: Any]?) -> ThemeOptions? {
        guard let theme = theme,
              let primaryColorValue = theme["primaryColor"] as? Int else {
            return nil
        }

        guard let primaryColor = color(from: primaryColorValue) else {
            return nil
        }
        let foreground = color(from: theme["foregroundOnPrimaryColor"])

        return FlutterThemeOptions(
            primary: primaryColor,
            primaryTint10: primaryColor,
            primaryTint20: primaryColor,
            primaryTint30: primaryColor,
            foregroundOnPrimary: foreground ?? .white
        )
    }

    private func buildLocalizationOptions(_ localization: [String: Any]?) -> LocalizationOptions? {
        guard let localization = localization,
              let languageCode = localization["languageCode"] as? String else {
            return nil
        }

        let locale: Locale
        if let countryCode = localization["countryCode"] as? String, !countryCode.isEmpty {
            locale = Locale(identifier: "\(languageCode)_\(countryCode)")
        } else {
            locale = Locale(identifier: languageCode)
        }

        let isRtl = localization["isRightToLeft"] as? Bool ?? false
        return LocalizationOptions(
            locale: locale,
            layoutDirection: isRtl ? .rightToLeft : .leftToRight
        )
    }

    private func buildOrientationOption(_ orientation: String?) -> OrientationOptions? {
        guard let orientation = orientation else { return nil }

        switch orientation {
        case "portrait":
            return .portrait
        case "landscape":
            return .landscape
        case "landscapeRight":
            return .landscapeRight
        case "landscapeLeft":
            return .landscapeLeft
        case "allButUpsideDown":
            return .allButUpsideDown
        default:
            return nil
        }
    }

    private func setupEventHandlers(composite: CallComposite) {
        composite.events.onError = { [weak self] errorEvent in
            let errorCode: String
            switch errorEvent.code {
            case CallCompositeErrorCode.callJoin:
                errorCode = "callJoinFailed"
            case CallCompositeErrorCode.callEnd:
                errorCode = "callEndFailed"
            case CallCompositeErrorCode.tokenExpired:
                errorCode = "tokenExpired"
            case CallCompositeErrorCode.cameraFailure:
                errorCode = "cameraFailure"
            case CallCompositeErrorCode.microphonePermissionNotGranted:
                errorCode = "microphonePermissionNotGranted"
            case CallCompositeErrorCode.networkConnectionNotAvailable:
                errorCode = "networkConnectionNotAvailable"
            case "callEvicted":
                errorCode = "callEvicted"
            case "callDeclined":
                errorCode = "callDeclined"
            default:
                errorCode = "unknown"
            }

            let nsError = errorEvent.error as NSError?
            let nativeMessage = nsError?.localizedDescription ?? errorEvent.error?.localizedDescription
            #if DEBUG
            let errorDetails = nsError != nil ? " domain=\(nsError!.domain) code=\(nsError!.code)" : ""
            NSLog("[AcsUiLibraryPlugin] onError code=\(errorCode) nativeCode=\(errorEvent.code) message=\(nativeMessage ?? "nil")\(errorDetails)")
            #endif

            DispatchQueue.main.async {
                self?.channel?.invokeMethod("onError", arguments: [
                    "errorCode": errorCode,
                    "message": nativeMessage,
                    "nativeCode": "\(errorEvent.code)",
                    "nativeDomain": nsError?.domain,
                    "nativeErrorCode": nsError?.code
                ])
            }
        }

        composite.events.onCallStateChanged = { [weak self] callStateEvent in
            let state: String
            switch callStateEvent.requestString {
            case "none":
                state = "none"
            case "connecting":
                state = "connecting"
            case "ringing":
                state = "ringing"
            case "connected":
                state = "connected"
            case "localHold":
                state = "localHold"
            case "remoteHold":
                state = "remoteHold"
            case "disconnecting":
                state = "disconnecting"
            case "disconnected":
                state = "disconnected"
            case "inLobby":
                state = "inLobby"
            default:
                state = "unknown"
            }

            #if DEBUG
            NSLog("[AcsUiLibraryPlugin] callStateChanged nativeState=\(callStateEvent.requestString) mapped=\(state) "
                  + "endCode=\(callStateEvent.callEndReasonCodeInt ?? -1) "
                  + "endSubCode=\(callStateEvent.callEndReasonSubCodeInt ?? -1) "
                  + "callId=\(callStateEvent.callId ?? "nil")")
            #endif

            DispatchQueue.main.async {
                self?.channel?.invokeMethod("onCallStateChanged", arguments: [
                    "state": state,
                    "nativeState": callStateEvent.requestString,
                    "callEndReasonCode": callStateEvent.callEndReasonCodeInt,
                    "callEndReasonSubCode": callStateEvent.callEndReasonSubCodeInt,
                    "callId": callStateEvent.callId
                ])
            }
        }

        composite.events.onDismissed = { [weak self] dismissedEvent in
            var args: [String: Any?] = [:]
            if let errorCode = dismissedEvent.errorCode {
                args["errorCode"] = "\(errorCode)"
            }
            let reason = dismissedEvent.error?.localizedDescription
            args["reason"] = reason

            #if DEBUG
            NSLog("[AcsUiLibraryPlugin] dismissed errorCode=\(args["errorCode"] ?? "nil") reason=\(reason ?? "nil")")
            #endif

            DispatchQueue.main.async {
                self?.channel?.invokeMethod("onDismissed", arguments: args)
                self?.callComposite = nil
            }
        }

        composite.events.onRemoteParticipantJoined = { [weak self] participantEvent in
            let ids = participantEvent.map { $0.rawId }
            DispatchQueue.main.async {
                self?.channel?.invokeMethod("onRemoteParticipantJoined", arguments: [
                    "participantCount": ids.count,
                    "participantIds": ids
                ])
            }
        }
    }

    private func color(from value: Any?) -> UIColor? {
        guard let colorValue = value as? Int else { return nil }

        let alpha = CGFloat((colorValue >> 24) & 0xFF) / 255.0
        let red = CGFloat((colorValue >> 16) & 0xFF) / 255.0
        let green = CGFloat((colorValue >> 8) & 0xFF) / 255.0
        let blue = CGFloat(colorValue & 0xFF) / 255.0

        return UIColor(red: red, green: green, blue: blue, alpha: alpha)
    }
}

private struct FlutterThemeOptions: ThemeOptions {
    let primaryColor: UIColor
    let primaryColorTint10: UIColor
    let primaryColorTint20: UIColor
    let primaryColorTint30: UIColor
    let foregroundOnPrimaryColor: UIColor

    var colorSchemeOverride: UIUserInterfaceStyle {
        return .unspecified
    }

    init(primary: UIColor,
         primaryTint10: UIColor,
         primaryTint20: UIColor,
         primaryTint30: UIColor,
         foregroundOnPrimary: UIColor) {
        self.primaryColor = primary
        self.primaryColorTint10 = primaryTint10
        self.primaryColorTint20 = primaryTint20
        self.primaryColorTint30 = primaryTint30
        self.foregroundOnPrimaryColor = foregroundOnPrimary
    }
}
